Allow whitelist of packages that can exit

jasontedor opened the pull request:
Today, SecureSM has a mechanism that enables a hardcoded list of test
packages to exit if the SecureSM instance is constructed with a boolean
flag indicating that these packages will be permitted to exit. This
commit replaces this mechanism by allowing the SecureSM instance to be
constructed with a whitelist of packages that can exit.

jasontedor committed:
Allow whitelist of packages that can exit

Today, SecureSM has a mechanism that enables a hardcoded list of test
packages to exit if the SecureSM instance is constructed with a boolean
flag indicating that these packages will be permitted to exit. This
commit replaces this mechanism by allowing the SecureSM instance to be
constructed with a whitelist of packages that can exit.

rmuir commented on src/main/java/org/elasticsearch/SecureSM.java:
```diff
@@ -170,28 +162,13 @@ protected void checkThreadGroupAccess(ThreadGroup g) {
    * The "Uwe Schindler" algorithm.
    */
-  protected void checkTestExit(final int status) {
+  protected void checkExit(final int status, final String[] packagesThatCanExit) {
```
I don't understand why we should change this signature, given that we already have the list of packages as an instance variable.

jasontedor replied:
I pushed f0dc394a62d4af190bb36d3ffd7f7a10d8a955dc.

rmuir commented on src/main/java/org/elasticsearch/SecureSM.java:
```diff
@@ -170,28 +162,13 @@ protected void checkThreadGroupAccess(ThreadGroup g) {
-    "org.eclipse.jdt.internal.junit.runner.",
-    // intellij test runner
-    "com.intellij.rt.execution.junit."
-  };
```
I also think the ctor should default to this list. why make someone figure it out?

jasontedor replied:
I think that it would be odd that the default constructor would allow test packages to exit instead of no packages to exit? I think we can expose these test packages as a convenience field though?

rmuir replied:
How about a static method instead, so it can return an array. I don't like a field because it means pulling in java.util.List and other things. Frankly. we shouldn't even be pulling in java.util.Objects as-is. We have to be (more) minimal to protect ourselves against bugs like https://bugs.openjdk.java.net/browse/JDK-8155090 and https://bugs.openjdk.java.net/browse/JDK-8155659

jasontedor replied:
I pushed 77f27fd4964ab1266ff64a9d04006d8121cee529.

jasontedor committed:
Simplify SecureSM#checkExit method signature

This commit removes a parameter from SecureSM#checkExit(int, String[])
that is not needed as the necessary values are a field on the instance.

jasontedor committed:
Provide method for creating test SecureSM

This commit adds a convenience method for creating a test SecureSM for
which a standard set of packages can exit or halt the VM. Additionally,
a default constructor is added with the semantics that no package can
exit nor halt the virtual machine.

rmuir commented on src/main/java/org/elasticsearch/SecureSM.java:
```diff
@@ -22,6 +22,9 @@
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
 import java.util.Objects;
```
can we nuke these imports when pushing?

jasontedor replied:
I pushed c8d058f47b5f39735bdc5e41cd84b6685dc20e90.

rmuir commented:
+1, thank you. i left a note about unused imports, looks great.

jasontedor committed:
Remove unused imports from SecureSM.java

jasontedor commented:
Thanks @rmuir.

rmuir commented:
now to figure out how to release a new version :)
